railgun/test/performance/perf_ingress_test.go

```diff
@@ -0,0 +1,82 @@
+//+build performance_tests
+
+package performance
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kong/kubernetes-ingress-controller/pkg/annotations"
+	k8sgen "github.com/kong/kubernetes-testing-framework/pkg/generators/k8s"
+)
+
+func TestIngressPerformance(t *testing.T) {
+	t.Log("setting up the TestIngressPerf")
+
+	ctx := context.Background()
+	cnt := 1
+	cost := 0
+	for cnt <= max_ingress {
+		namespace := fmt.Sprintf("ingress-%d", cnt)
+		nsName := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: namespace,
+			},
+		}
+		_, err := cluster.Client().CoreV1().Namespaces().Create(context.Background(), nsName, metav1.CreateOptions{})
+		assert.NoError(t, err)
+
+		t.Log("deploying a minimal HTTP container deployment to test Ingress routes")
+		container := k8sgen.NewContainer("httpbin", httpBinImage, 80)
+		deployment := k8sgen.NewDeploymentForContainer(container)
+		deployment, err = cluster.Client().AppsV1().Deployments(namespace).Create(ctx, deployment, metav1.CreateOptions{})
+		assert.NoError(t, err)
+
+		t.Logf("exposing deployment %s via service", deployment.Name)
+		service := k8sgen.NewServiceForDeployment(deployment, corev1.ServiceTypeLoadBalancer)
+		_, err = cluster.Client().CoreV1().Services(namespace).Create(ctx, service, metav1.CreateOptions{})
+		assert.NoError(t, err)
+
+		t.Logf("[%s] creating an ingress for service httpbin with ingress.class %s", namespace, ingressClass)
+		ingress := k8sgen.NewIngressForService("/httpbin", map[string]string{
+			annotations.IngressClassKey: ingressClass,
+			"konghq.com/strip-path":     "true",
+		}, service)
+		ingress, err = cluster.Client().NetworkingV1().Ingresses(namespace).Create(ctx, ingress, metav1.CreateOptions{})
+		assert.NoError(t, err)
+		start_time := time.Now().Nanosecond()
+
+		t.Logf("checking routes from Ingress %s to be operational", ingress.Name)
+		assert.Eventually(t, func() bool {
+			resp, err := httpc.Get(fmt.Sprintf("%s/httpbin", KongInfo.ProxyURL.String()))
+			if err != nil {
+				t.Logf("WARNING: error while waiting for %s: %v", KongInfo.ProxyURL.String(), err)
+				return false
+			}
+			defer resp.Body.Close()
+			if resp.StatusCode == http.StatusOK {
+				// now that the ingress backend is routable, make sure the contents we're getting back are what we expect
+				// Expected: "<title>httpbin.org</title>"
+				b := new(bytes.Buffer)
+				b.ReadFrom(resp.Body)
+				end_time := time.Now().Nanosecond()
+				loop := end_time - start_time
+				t.Logf("networkingv1 hostname is ready to redirect traffic after %d.", loop)
+				cost += loop
+				return strings.Contains(b.String(), "<title>httpbin.org</title>")
+			}
+			return false
+		}, ingressWait, waitTick)
```

Comment:
the `waitTick` here is currently set to 1second, this is going to cause drag on our `cost` and can make it cost 1s+ when perhaps the actual resolution time is subsecond.

I think we need to make a decision about what kind of time resolution we're actually looking for with these tests, but perhaps something like 1ms tick time for this particular check would be OK, meaning we VERY quickly try again instead of waiting a whole second?

```suggestion
		}, ingressWait, time.Millisecond * 1)
```

Reply:
> Ideally I was hoping any performance tooling we add would be something we could trigger with a CLI, or perhaps a docker container, e.t.c. This mechanism will lock performance testing tightly into our testing framework, which does not seem to align well with some of the stated goals in #1197.
> 
> Particularly, the main goal stated there:
> 
> ```
> We want to create a performance testing tool that can be applied generically with any particular Kubernetes cluster backend.
> ```
> 
> Building the performance testing into our KIC tests will _specifically not be generic_ as it will mean:
> 
> 1. anyone wishing to run the performance tests needs to be educated on our git repo
> 2. anyone wishing to run the performance tests needs to understand `go test`
> 
> And also we currently only support `kind` on the `github actions` infrastructure, which not only doesn't seem like a strong testing environment (I believe this environment is highly contentious for compute resources being on shared infrastructure) but also doesn't allow for any other cluster options as per the `any particular Kubernetes cluster backend` bit.
> 
> My recommendation would be for us to transplant the functionality that is currently in `railgun/test/integration` into a discreet library which could then be consumed by the integration tests for basic validation and regression testing, but also then can be consumed more generally by some CLI or other caller who can bring whatever Kubernetes cluster (via `*rest.Config`) to the table they want to.

I like the idea a docker container dedicated for performance tests. As performance tests requests extra cpu/memory/network resource; also, controller manager need scale to satisfy the request,  we can't mix performance tests with other CI ongoing tests for an accurate test results. 

I don't think we need re-invent the wheel, just leverage the test cases under performance directory and collect data from there. 
based on this thought, revert back to WIP.